src/gc/gc.d

```diff
@@ -2759,19 +2763,21 @@ struct Gcx
                 Lnotfree:
                     p = pool.baseAddr + pn * PAGESIZE;
                     for (u = 0; u < PAGESIZE; u += size)
-                    {   biti = bitbase + u / 16;
-                        if (pool.freebits.test(biti))
-                        {
-                            List *list = cast(List *)(p + u);
-                            if (list.next != bucket[bin])       // avoid unnecessary writes
-                                list.next = bucket[bin];
-                            list.pool = pool;
-                            bucket[bin] = list;
-                        }
+                    {
+                        biti = bitbase + u / 16;
+                        if (!pool.freebits.test(biti))
+                            continue;
+                        auto elem = cast(List *)(p + u);
+                        elem.pool = pool;
+                        *tail[bin] = elem;
+                        tail[bin] = &elem.next;
                     }
```

Comment:
Have you tried keeping the update of bucket[bin] local to this loop? Maybe this pays off when optimizing the loop.